### PR TITLE
STN-692 : Create Vertical Timeline Component

### PR DESCRIPTION
**Pull request by heatherdesigns**
# [STN-692](https://sparkbox.atlassian.net/browse/STN-692) READY FOR REVIEW

## Summary
The work in this PR creates a Vertical Timeline component that allows for multiple “timeline item” elements to be styled together so that presents a step-by-step content flow in a clear format to users.

_Note:_ The vertical timelines is not currently scoped to work with dark color inversion.

## Need Review By (Date)
asap

## Urgency
hot 🔥 

## Steps to Test
1. In your CLI, run `npm run test` and confirm all tests pass.
2. In order for the vertical timeline and timeline patterns to be recognized, you will need to clear your cache by running `lando drush @sparkbox_sandbox.local cr`. Run `lando drush @sparkbox_sandbox.local config-import --partial` to import the config files.
3. To test a vertical timeline, create a new flexible page or open an existing one. Under Components add a Vertical Timeline. Fill in all the fields.

Test the following scenarios:
- [x] A vertical timeline with the Add Expand/Collapse checkbox NOT checked. The timeline should be expanded by default on page load. When the timeline is expanded, there should not be an arrow icon to the right of the timeline items.
- [x] A vertical timeline with the Add Expand/Collapse checkbox checked. The timeline should be collapsed. There should be an arrow icon visible to the right of the timeline item.
- [x] A vertical timeline with a single timeline item should only have the circle icon to the left. It should not have a vertical pipe connected to it.
- [x] A vertical timeline in a row.
- [x] A vertical timeline in a well.
- [x] A vertical timeline in a row beside another component such as a text area.
- [x] A vertical timeline with a timeline item that only has a summary. Confirm this is the only required field.
- [x] A vertical timeline with a timeline item that has a summary and subheading only.
- [x] A vertical timeline with a timeline item that has a summary, subheading, and description.
- [x] Test both Traditional and Colorful themes. The only difference you should see is the line-height of the description. In Colorful the line height is a little more generous.

4. Confirm the timeline item pattern is not available as a paragraph in a row or in the components field on a flexible page.

### CODE REVIEW

**Comment by rkelbel** [comment]
Heather this work is great and looks **soooo good**! I have a couple of config updates for you here that may be useful for user experience.

I noticed Timeline Item is available in the Components area and within a Row Component area to be added without a Timeline Component, I don't think this was intended.

**Flexible Page Components Area:**

This may be because a config file is missing from the export, the file that gets edited for this is:
`field.field.node.hs_basic_page.field_hs_page_components`
  
You can set to exclude the Timeline Item from the Components field on Flexible/Basic Pages by going here: http://sparkbox-sandbox.suhumsci.loc/admin/structure/types/manage/hs_basic_page/fields and clicking Edit on Components field, then checking any item you would like to exclude.

---

**To update this for the Row options:**

The file you'll want to update is:
`field.field.paragraph.hs_row.field_hs_row_components`
  
You can update the components field within Rows here: http://sparkbox-sandbox.suhumsci.loc/admin/structure/paragraphs_type/hs_row/fields you'll want to check anything you want to exclude as you do with the Flexible/Basic page setup as well.

Once those are done you should be able to export that file for the field so that is inherited upon `Config:import` on the receiving end.

**Comment by rkelbel** [comment]
**Timelines within Wells:**
Should we add some padding to the top of the timeline when they are added to Rows?  This would give them some more breathing room at the top and let them look centered within the Row if we added the same spacing to the top that is added to the bottom of a Timeline item.

This applies to both Traditional and Colorful:
image

image

File: docroot/themes/humsci/humsci_basic/src/scss/objects/_layouts.general.scss [hunk 1]
```diff
@@ -35,6 +35,7 @@
 .ptype-hs-carousel,
 .field-hs-page-hero,
 .ptype-hs-row,
+.ptype-hs-timeline,
```
Comment (heatherdesigns) [inline docroot/themes/humsci/humsci_basic/src/scss/objects/_layouts.general.scss:38]: I'm not seeing the same issue with the margin-top when a timeline is in a row with a well. This class _should_ be taking care of that. Let me know if you still see the issue so we can pair and I can see what you see in your local environment. 💃 

well

Comment (rkelbel) [reply docroot/themes/humsci/humsci_basic/src/scss/objects/_layouts.general.scss:38]: Yep, I think this was browser caching the padding on this, it is indeed fine. Thank you Heather!

File: config/default/field.field.node.hs_basic_page.field_hs_page_components.yml [hunk 2]
```diff
@@ -42,6 +44,9 @@ settings:
       hs_carousel:
         weight: 15
         enabled: false
+      hs_gradient_hero:
+        weight: 21
+        enabled: false
```
Comment (rkelbel) [inline config/default/field.field.node.hs_basic_page.field_hs_page_components.yml]: This configuration is present for hs_gradient_hero.  We probably don't want this to be showing up in here yet.

File: config/default/field.field.paragraph.hs_row.field_hs_row_components.yml [hunk 3]
```diff
@@ -41,6 +43,9 @@ settings:
       hs_carousel:
         weight: 12
         enabled: false
+      hs_gradient_hero:
+        weight: 21
+        enabled: false
```
Comment (rkelbel) [inline config/default/field.field.paragraph.hs_row.field_hs_row_components.yml]: Also config for hs_gradient_hero in this file.

File: docroot/themes/humsci/humsci_basic/src/scss/components/_pattern.timeline-item.scss [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  &__revealed-details {
+    padding-bottom: hb-spacing-width('sm');
+
+    // only animate collapsed timeline items when they are opened
+    // timelines that are expanded by default should not animate
+    .hb-timeline-item[open] & {
+      .hb-timeline__collapsed & {
+        animation: timelineFadeInDown 350ms ease-in-out forwards;
```
Comment (rkelbel) [inline docroot/themes/humsci/humsci_basic/src/scss/components/_pattern.timeline-item.scss:60]: This is nice!

File: docroot/themes/humsci/humsci_basic/patterns/timeline/timeline.ui_patterns.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+      type: text
+      preview: ""
+  variants:
+    expanded:
+       label: Expanded (Default)
+       description: All timeline items are expanded (open) on page load.
+    collapsed:
+       label: Collapsed
+       description: All timeline items are collapsed (closed) on page load.
+  use: "@humsci_basic/timeline/timeline.html.twig"
```
Comment (rkelbel) [inline docroot/themes/humsci/humsci_basic/patterns/timeline/timeline.ui_patterns.yml]: Can you add a new line to the bottom of this file please?

File: docroot/themes/humsci/humsci_basic/src/js/shared/index.js [hunk 6]
```diff
@@ -5,4 +5,5 @@ import './tables/table-pattern';
 import './navigation/index';
 import './equal-height-grid/index';
 import './media/video-with-caption';
-import './animation/page-scroll';
\ No newline at end of file
+import './animation/page-scroll';
+import './timeline/expand-collapse-timeline';
```
Comment (rkelbel) [inline docroot/themes/humsci/humsci_basic/src/js/shared/index.js]: We should also probably add a new line to the bottom of this file as well.

File: docroot/themes/humsci/humsci_basic/src/scss/components/_pattern.timeline-item.scss [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  0% {
+    opacity: 0;
+    transform: translateY(-8px);
+  }
+
+  100% {
+    opacity: 1;
+    transform: translateY(0);
+  }
+}
```
Comment (rkelbel) [inline docroot/themes/humsci/humsci_basic/src/scss/components/_pattern.timeline-item.scss:139]: This component looks and works so great Heather!  I can't wait to see it being used on sites! ❤️

**Comment by rkelbel** [comment]
Nice work Heather!

high five

**Pull request merged**